src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+        let mut rdr = Cursor::new(payload);
+        let accepted = rdr
+            .read_u8()
+            .map_err(|_e| "Controller version reply too short")?;
+        Ok(accepted != 0)
+    }
+
+    fn encode_configure_inputs_request<U: AsRef<str>>(variables: &[U]) -> Result<Vec<u8>, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        if !variables.iter().all(|v| !v.contains(',')) {
+            return Err(format!(
+                "Comma can't be part of variable name in {:?}",
+                variables
+            ));
+        }
+        let payload_as_string = variables.join(",");
+        let payload = payload_as_string.as_bytes().to_vec();
+        Self::encode_packet(RTDECommand::ConfigureInputs as u8, payload)
+    }
+
+    fn encode_configure_outputs_request<U: AsRef<str>>(
+        frequency: f64,
+        variables: &[U],
+    ) -> Result<Vec<u8>, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        if !variables.iter().all(|v| !v.contains(',')) {
+            return Err(format!(
+                "Comma can't be part of variable name in {:?}",
+                variables
+            ));
+        }
+        let mut payload: Vec<u8> = vec![];
+        payload.write_f64::<BigEndian>(frequency).unwrap();
+        payload.extend(variables.join(",").as_bytes());
+
+        Self::encode_packet(RTDECommand::ConfigureOutputs as u8, payload)
+    }
+
+    fn map_type_name_to_discriminant(type_name: &str) -> Result<Discriminant<RTDEData>, String> {
+        let default_value = RTDEData::default_for_type_name(type_name)?;
+        Ok(discriminant(&default_value))
+    }
+
+    fn decode_configure_io_reply<U: AsRef<str>>(
+        payload: Vec<u8>,
+        variables: &[U],
+    ) -> Result<IoSetup, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        let mut rdr = Cursor::new(payload);
+        let recipe_id: u8 = rdr.read_u8().map_err(|_e| "Inputs setup reply too short")?;
+        let payload = rdr.into_inner();
+        let string = String::from_utf8(payload[1..].to_vec())
+            .map_err(|_e| format!("Inputs reply contains non-UTF8 chars: {:?}", &payload[1..]))?;
+        let type_names: Vec<String> = string.split(',').map(String::from).collect();
+
+        if let Some(offender) = type_names.iter().position(|t| t == "NOT_FOUND") {
+            return Err(format!("Variable {} not found", variables[offender]));
+        }
+        if let Some(offender) = type_names.iter().position(|t| t == "IN_USE") {
+            return Err(format!(
+                "Variable {} is already in use by another connection",
+                variables[offender]
+            ));
+        }
+
+        let types: Vec<Discriminant<RTDEData>> = type_names
+            .iter()
+            .map(|n| Self::map_type_name_to_discriminant(n))
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
You already have a type annotation on `types`. You could drop the `Vec<_>` from the turbofish of `.collect()`.

Reply:
The tricky part here is that `?` at the end, so we have a intermediary `Result`, which seems problematic for type inference. If I remove the turbofish, I get:
```
error[E0284]: type annotations needed
   --> ur_rtde/src/raw_rtde/connection.rs:274:14
    |
274 |             .collect()?;  // ::<Result<Vec<_>, _>>
    |              ^^^^^^^ cannot infer type
    |
    = note: cannot satisfy `<_ as Try>::Residual == _`
help: consider specifying the type argument in the method call
    |
274 |             .collect::<B>()?;  // ::<Result<Vec<_>, _>>
    |                     +++++

error[E0283]: type annotations needed
   --> ur_rtde/src/raw_rtde/connection.rs:274:23
    |
274 |             .collect()?;  // ::<Result<Vec<_>, _>>
    |                       ^ cannot infer type for type parameter `E`
    |
    = note: multiple `impl`s satisfying `String: ~const From<_>` found in the following crates: `alloc`, `core`:
            - impl From<Box<str>> for String;
            - impl From<char> for String;
            - impl<'a> From<Cow<'a, str>> for String;
            - impl<> From<&String> for String;
            and 3 more
    = note: required because of the requirements on the impl of `FromResidual<Result<Infallible, _>>` for `Result<packet_data::IoSetup, String>`

Some errors have detailed explanations: E0283, E0284.
For more information about an error, try `rustc --explain E0283`.

``` 
That's expected, or is there a way around it?

Reply:
I said drop the `Vec<_>`, not the *entire* type annotation.

Reply:
Right, just dropping the `Vec<_>` works.